BEPs/BEP-396.md

```diff
@@ -0,0 +1,187 @@
+
+There is also a potential `TxDAG` "compression algorithm" to further optimize the size. For the DAG describing the dependencies, some are redundant because it only needs to ensure that the final execution path is consistent.
+
+![](./assets/BEP-396/pic2.png)
+
+Specific rules: If Txj depends on Txi, j>i, then Txj can ignore the same dependencies as Txi, and the execution path of Txj after the update remains the same as before.
+
+This can keep the size of `TxDAG` as small as possible and avoid generating too large `TxDAG` and wasting network bandwidth.
+
+### 5.3 TxDAG Validation
+
+Since there is a preset order of transactions in the block, Txi can't depend on Txj (i<j) in DAG, and it is also impossible for circular dependencies to occur. 
+
+The verification rules are as follows:
+
+1. TxIndexes must be smaller than the current TxIndex;
+2. TxIndexes cannot be repeated;
+3. TxIndexes must be sorted in ascending order;
+4. TxIndex in `TxDAG` cannot exceed the maximum TxIndex of the block;
+5. Verify whether the dependency optimization rules are followed and delete redundant dependencies;
+
+### 5.4 Block Extension
+
+This BEP stores `TxDAG` data in the `BlockBody`, and its Hash in the Extra field of the `Header`. To ensure the correctness of the DAG, if the validator finds an incorrect DAG, it should reject the block.
+
+This is a diagram of the Extra field structure changes in the `Header`:
+
+![](assets/BEP-396/pic3.png)
+
+The specific `BlockBody` structure is defined as follows:
+
+```go
+type Body struct {
+	Transactions []*Transaction
+	Uncles       []*Header
+	Withdrawals  []*Withdrawal `rlp:"optional"`
+	TxDAG        *TxDAG        `rlp:"optional"`
+}
+```
+
+### 5.5 TxDAG Generation
+
+The generation of `TxDAG` requires analysis of the transaction read-write set. You can set a read-write tracker in the EVM's `StateDB`. When reading the state from the DB, it is recorded in the read set. When creating or updating the state, it is recorded in the write set. When reverting the modification, the write set also needs to be reverted.
+
+Finally, the read-write dependency of the transaction is traversed to generate `TxDAG`.
+
+### 5.6 Validating Block
+
+The easiest way to speed up validating blocks through `TxDAG` is to integrate with parallel processors. Usually, parallel processors only optimistically execute transactions in parallel. If `TxDAG` is input, the parallel processor can execute the entire block in the best parallel path. The parallel processor can easily find bad DAG.
+
+At the same time, `TxDAG` can be applied in state prefetcher scenarios, and the state can be loaded in advance through simple parallel prefetchers to improve the cache hit rate of the processor.
+
+![](assets/BEP-396/pic4.png)
+
+If the DAG generated locally by the validator does not match the block, this block should be rejected.
```

Comment:
It is possible that the validator generate a correct DAG, while the DAG processor of a fullnode get bugs, if fullnode just reject it, then the whole network will halt, I will suggest to fall back to the common processor without parallel in such case.

Reply:
Yeah, there may be some tough bugs in the parallel processor, and it's also important. But it may not be part of BEP, this BEP aims to add TxDAG to accelerate block execution and keep it correct too(serial or parallel processors both can produce & verify TxDAG).

At the same time, grayscale testing, cross-validation, or candidates are not recommended to enable parallel processors are all early stages to avoid network-wide failures. Of course, node diversity and the multiple processors in the network are more reasonable and more conducive to stability and security.

Reply:
let us take care of the first implementation version